admin/javascript/src/components/form-builder/index.js

```diff
@@ -0,0 +1,229 @@
+        return null;
+      case 'HTML':
+        // HTML editor field (not implemented)
+        return null;
+      case 'Integer':
+        // Numeric field (not implemented)
+        return null;
+      case 'Decimal':
+        // Numeric field (not implemented)
+        return null;
+      case 'MultiSelect':
+        // Radio field (not implemented)
+        return null;
+      case 'SingleSelect':
+        // Dropdown field (not implemented)
+        return null;
+      case 'Date':
+        // DateTime field (not implemented)
+        return null;
+      case 'DateTime':
+        // DateTime field (not implemented)
+        return null;
+      case 'Time':
+        // DateTime field (not implemented)
+        return null;
+      case 'Boolean':
+        // Checkbox field (not implemented)
+        return null;
+      case 'Custom':
+        return this.components.GridField;
+      default:
+        return null;
+    }
+  },
+};
+
+export class FormBuilderComponent extends SilverStripeComponent {
+
+  constructor(props) {
+    super(props);
+
+    this.formSchemaPromise = null;
+    this.isFetching = false;
+
+    this.fetch();
+  }
+
+  /**
+   * Fetches data used to generate a form. This can be form schema and or form state data.
+   * When the response comes back the data is saved to state.
+   *
+   * @param boolean schema - If form schema data should be returned in the response.
+   * @param boolean state - If form state data should be returned in the response.
+   *
+   * @return object - Promise from the AJAX request.
+   */
+  fetch(schema = true, state = false) {
+    const headerValues = [];
+
+    if (this.isFetching === true) {
+      return this.formSchemaPromise;
+    }
+
+    if (schema === true) {
+      headerValues.push('schema');
+    }
+
+    if (state === true) {
+      headerValues.push('state');
+    }
+
+    this.formSchemaPromise = fetch(this.props.schemaUrl, {
+      headers: { 'X-FormSchema-Request': headerValues.join() },
+      credentials: 'same-origin',
+    })
+      .then(response => response.json())
+      .then(json => {
+        this.isFetching = false;
+        this.props.actions.setSchema(json);
+      });
+
+    this.isFetching = true;
+
+    return this.formSchemaPromise;
+  }
+
+  /**
+   * Maps a list of schema fields to their React Component.
+   * Only top level form fields are handled here, composite fields (TabSets etc),
+   * are responsible for mapping and rendering their children.
+   *
+   * @param array fields
+   *
+   * @return array
+   */
+  mapFieldsToComponents(fields) {
+    return fields.map((field, i) => {
+      const Component = field.component !== null
+        ? fakeInjector.getComponentByName(field.component)
+        : fakeInjector.getComponentByDataType(field.type);
+
+      if (Component === null) {
+        return null;
+      }
+
+      // Props which every form field receives.
+      const props = {
+        attributes: field.attributes,
```

Comment:
`attributes: {...field.attributes}` isn't actually part of the ES6 spec yet, I had to enable "Stage 2: Draft" in Babel for it to parse (http://babeljs.io/docs/plugins/#presets). And even then, I can't get ESLint to parse it. So in short, it's probably a bit too early to use the syntax like that. I've resorted to the `deep-freeze` library we already have in the codebase for other purposes.